src/lib.rs

```diff
@@ -16,12 +16,44 @@ pub trait Gcd {
     /// assert_eq!(44, 2024u32.gcd(748));
     /// ```
     fn gcd(self, other: Self) -> Self;
+    fn gcd_binary(self, other: Self) -> Self;
+    fn gcd_euclid(self, other: Self) -> Self;
 }
 
 macro_rules! gcd_impl {
     ($($t:ty),*) => ($(
         impl Gcd for $t {
-            fn gcd(self, other: Self) -> Self {
+            fn gcd(self,other: Self) -> Self {
+                self.gcd_binary(other)
+            }
+            fn gcd_binary(self, mut v: Self) -> Self {
+                let mut u = self;
+                if u == 0 {
+                    return v;
+                }
+                if v == 0 {
+                    return u;
+                }
+                let shift = (u | v).trailing_zeros();
+                u >>= shift;
+                v >>= shift;
+                u >>= u.trailing_zeros();
+                loop {
+                    v = v >> (v.trailing_zeros());
```

Comment:
This could use the same format as the above call for `u`. Just for consistency.

```suggestion
                    v >>= v.trailing_zeros();
```